internal/graph/cycle_check_resolver_test.go

```diff
@@ -0,0 +1,83 @@
+package graph
+
+import (
+	"context"
+	"testing"
+
+	"github.com/oklog/ulid/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/storage/memory"
+	"github.com/openfga/openfga/pkg/testutils"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+func TestCycleDetectionCheckResolver(t *testing.T) {
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+	})
+
+	ds := memory.New()
+	t.Cleanup(ds.Close)
+	storeID := ulid.Make().String()
+
+	err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+		tuple.NewTupleKey("document:1", "viewer3", "document:1#viewer3"),
+		tuple.NewTupleKey("document:1", "viewer1", "user:1"),
+	})
+	require.NoError(t, err)
+
+	model := testutils.MustTransformDSLToProtoWithID(`model
+	schema 1.1
+  
+  type user
+  
+  type document
+	relations
+		define viewer1: [user, document#viewer1]
+		define viewer2: viewer1 or viewer2
```

Comment:
This is an invalid model as well. I'm surprised as hell `MustTransformDSLToProtoWithID` is not throwing an error 😕 

cc'ing @rhamzeh on this.

We shouldn't be able to define `viewer2: viewer1 or viewer2`, because `viewer2` is not assignable. Playground rejects this today, and the server rejects it as well.

See:
<img width="1467" alt="Screenshot 2024-02-27 at 3 53 20 PM" src="https://github.com/openfga/openfga/assets/2899204/2ef6ec7b-7ef9-49a0-a82c-64b15b544af2">
and the following request against the server fails:
```
curl --location 'http://localhost:8080/stores/<store_id>/authorization-models' \
--header 'Content-Type: application/json' \
--data '{
  "schema_version": "1.1",
  "type_definitions": [
    {
      "type": "user",
      "relations": {},
      "metadata": null
    },
    {
      "type": "document",
      "relations": {
        "viewer1": {
            "this": {}
        },
        "viewer2": {
          "union": {
            "child": [
              {
                "computedUserset": {
                  "relation": "viewer1"
                }
              },
              {
                "this": {}
              }
            ]
          }
        },
        "viewer3": {
          "union": {
            "child": [
              {
                "computedUserset": {
                  "relation": "viewer1"
                }
              },
              {
                "computedUserset": {
                  "relation": "viewer2"
                }
              }
            ]
          }
        }
      },
      "metadata": {
        "relations": {
          "viewer1": {
            "directly_related_user_types": [
              {
                "type": "user"
              },
              {
                "type": "document",
                "relation": "viewer1"
              }
            ]
          }
        }
      }
    }
  ]
}'
```

Reply:
We have several instances of invalid models in the tests. In theory an "old" model could exist prior to cycle detection. 

The now-invalid model is actually intentional. The test just needs to create a cyclical situation, how it does that is not that important IMO. Now that we have more robust cycle detection, a now-valid model will need to be more complex and more difficult to reason about.